idx/memory/memory.go

```diff
@@ -82,63 +90,78 @@ func (m *MemoryIdx) Stop() {
 	return
 }
 
-func (m *MemoryIdx) Add(data *schema.MetricData, partition int32) error {
+func (m *MemoryIdx) AddOrUpdate(data *schema.MetricData, partition int32) error {
 	pre := time.Now()
 	m.Lock()
 	defer m.Unlock()
-	err, ok := m.FailedDefs[data.Id]
+	err, ok := m.FailedAdds[data.Id]
 	if ok {
 		// if it failed before, it would fail again.
 		// there's not much point in doing the work of trying over
 		// and over again, and flooding the logs with the same failure.
 		// so just trigger the stats metric as if we tried again
-		idxFail.Inc()
+		statAddFail.Inc()
 		return err
 	}
 	existing, ok := m.DefById[data.Id]
 	if ok {
 		log.Debug("metricDef with id %s already in index.", data.Id)
 		existing.LastUpdate = data.Time
-		idxOk.Inc()
-		idxAddDuration.Value(time.Since(pre))
+		statUpdateOk.Inc()
+		statUpdateDuration.Value(time.Since(pre))
 		return nil
 	}
 
 	def := schema.MetricDefinitionFromMetricData(data)
 	err = m.add(def)
-	idxAddDuration.Value(time.Since(pre))
+	if err == nil {
+		statMetricsActive.Inc()
+	}
+	statAddDuration.Value(time.Since(pre))
 	return err
 }
 
 // Used to rebuild the index from an existing set of metricDefinitions.
-func (m *MemoryIdx) Load(defs []schema.MetricDefinition) {
+func (m *MemoryIdx) Load(defs []schema.MetricDefinition) (int, error) {
 	m.Lock()
 	var pre time.Time
+	var num int
+	var firstErr error
 	for i := range defs {
 		def := defs[i]
 		pre = time.Now()
 		if _, ok := m.DefById[def.Id]; ok {
 			continue
 		}
-		m.add(&def)
-		idxAddDuration.Value(time.Since(pre))
+		err := m.add(&def)
```

Comment:
yes, but then you'd have an extra pointer dereference / memory lookup. not sure which is best.
we could probably optimize this but that's out of scope of this PR and i don't have much time now to give this the attention to detail it needs, it's also rather low prio.